### ISSUE TITLE
Parameters to change early stopping metric

### ISSUE THREAD

**Issue by Suniljit** [ultralytics/ultralytics#7535]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

Hi, according to issue #5925, we should be able to change the early stop metric from mAP to something else for YOLOv8 under the EarlyStopping class in torch_utils.py. 

Based on the extract below, self.best_fitness = 0.0 sets the metric to mAP. Could I check what other parameter values are available? For e.g., what value should I set to use validation loss instead? Seems like this is not covered in the documentation. Thanks. 

`class EarlyStopping:
    """Early stopping class that stops training when a specified number of epochs have passed without improvement."""

    def __init__(self, patience=50):
        """
        Initialize early stopping object.

        Args:
            patience (int, optional): Number of epochs to wait after fitness stops improving before stopping.
        """
        self.best_fitness = 0.0  # i.e. mAP
        self.best_epoch = 0
        self.patience = patience or float('inf')  # epochs to wait after fitness stops improving to stop
        self.possible_stop = False  # possible stop may occur next epoch
`

**Comment by glenn-jocher** [comment]
@Suniljit hello! The `best_fitness` attribute in the `EarlyStopping` class is initialized to `0.0`, but it's not directly tied to mAP or any specific metric. Instead, it's a placeholder for the best value of the metric you're monitoring. To change the early stopping metric to something like validation loss, you would need to modify the code where the early stopping condition is checked, and update the `best_fitness` variable based on your desired metric.

Currently, the documentation does not list all possible metrics you can use for early stopping, but common choices include validation loss, precision, recall, or any other metric that you find suitable for your training process. You would typically set the `best_fitness` to the best (lowest for loss, highest for performance metrics) observed value of your chosen metric during training.

If you need further guidance on implementing this, please refer to the documentation on the Train mode, which may provide additional context on how metrics are handled during the training process. 📘🚀

**Comment by Suniljit** [comment]
Hi Glenn, could I check what is the default early stop metric used for YOLOv8 and where do I change it? From what I could trace in the [trainer.py](https://github.com/ultralytics/ultralytics/blob/30185e0d4c5708677b769a6c42a55eeeea2a2ce3/ultralytics/engine/trainer.py#L547) file, seems like if no metric is found, val loss is used. But I can’t seem to find where the stopping metric is defined. Thanks! 

`fitness = metrics.pop("fitness", -self.loss.detach().cpu().numpy())  # use loss as fitness measure if not found`

**Comment by glenn-jocher** [comment]
@Suniljit hi there! The default early stop metric in YOLOv8 is indeed the validation loss if no other metric is specified. The line you're referring to in `trainer.py` is where the early stopping metric is determined. If a 'fitness' metric is not found in the metrics dictionary, it falls back to using the negative validation loss.

To change the early stopping metric, you would adjust the code where the 'fitness' value is set. This involves modifying the logic within the training loop where early stopping conditions are evaluated. The documentation on the Train mode provides more details on how training metrics are managed. 📘🚀
